### PR TITLE
Add `step.invoke()` support for SDKs

### PR DESCRIPTION
**Pull request by jpwilliams**
## Description

Adds support for `step.invoke()` to SDKs.

- Adds new `inngest/function.completed` internal event
- Adds new `StepInvokingFunction` op code and generator handler
- Adds new `OnInvokeFunction` and `OnInvokeFunctionResumed` lifecycle methods
- Adds new `InvokeFunction` and `InvokeFunctionResult` fields to history
- Refactors `event.data._inngest_fn` to be `event.data._inngest.fn_id` to make room for other such Inngest data

```mermaid
sequenceDiagram
autonumber

participant SDK
participant E as Executor
participant EI as Event PubSub

E->>SDK: Executes function
SDK-->>+E: 206 - StepInvokingFunction
E->>E: Create waitForEvent-style pause + timeout job<br/>for inngest/function.completed
E->>-EI: Send inngest/function.invoked event
EI-)+E: inngest/function.invoked
E->>E: Find function to invoke from ID in event
E->>-E: Initialize function...
note over SDK,EI: On function completion...
E->>+E: OnFunctionFinished lifecycle
E->>-EI: inngest/function.completed
EI-)+E: inngest/function.completed
E->>-E: Consume pause for inngest/function.completed,<br/>set data to result (in event body)
E->>SDK: Continue function
```

## Todo

- [x] Add toggle for internal `inngest/*` events that haven't triggered anything on dev server
- [x] Fix UI not displaying payloads in some cases
- [x] Failure cases + tests
  - [x] Function not found
  - [x] Function failed
- [x] TODOs

## Type of change (choose one)
- [ ] Chore (refactors, upgrades, etc.)
- [ ] Bug fix (non-breaking change that fixes an issue)
- [ ] Security fix (non-breaking change that fixes a potential vulnerability)
- [ ] Docs
- [x] New feature (non-breaking change that adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality not to work as expected)

## Checklist
- [x] I've linked any associated issues to this PR.
- [x] I've tested my own changes.

## Related

- inngest/inngest-js#368
- inngest/monorepo#1988

*[Check our Pull Request Guidelines](https://github.com/inngest/inngest/blob/main/docs/PULL_REQUEST_GUIDELINES.md)*

### CODE REVIEW

File: pkg/consts/consts.go [hunk 1]
```diff
@@ -64,12 +64,14 @@ const (
 	// MaxEvents is the maximum number of events we can parse in a single batch.
 	MaxEvents = 5_000
 
+	InngestEventDataPrefix = "_inngest"
```
Comment (jpwilliams) [inline pkg/consts/consts.go:67]: We already seemed to need two variables here, so wanted to roll them up. Either this or the `inngest/function.invoked` event namespaces `data` instead of us adding some `_inngest*` keys to it.

File: pkg/devserver/lifecycle.go [hunk 2]
```diff
@@ -45,3 +53,60 @@ func (l lifecycle) OnFunctionScheduled(
...
+	} else {
+		data["result"] = resp.Output
+	}
+
+	evt := event.Event{
+		ID:        ulid.MustNew(uint64(now.UnixMilli()), rand.Reader).String(),
+		Name:      event.FnCompletedName,
+		Timestamp: now.UnixMilli(),
+		Data:      data,
+	}
```
Comment (jpwilliams) [inline pkg/devserver/lifecycle.go:93]: Should this use the same `user` as the original event? To get a trail of all events stemming from a particular user this seems reasonable, though this isn't _triggered_ by the user.

File: pkg/devserver/lifecycle.go [hunk 3]
```diff
@@ -45,3 +53,60 @@ func (l lifecycle) OnFunctionScheduled(
...
+
+	_ = l.pb.Publish(
+		ctx,
+		l.eventTopic,
+		pubsub.Message{
+			Name:      event.EventReceivedName,
+			Data:      string(byt),
+			Timestamp: now,
+		},
+	)
```
Comment (jpwilliams) [inline pkg/devserver/lifecycle.go:111]: Sending this is pretty critical, though lifecycle methods probably shouldn't contain crticial flows. Thinking this should move outside of a lifecycle method.

File: pkg/event/event.go [hunk 4]
```diff
@@ -7,12 +7,14 @@ import (
...
 	"time"
 
+	"github.com/inngest/inngest/pkg/consts"
 	"github.com/oklog/ulid/v2"
 )
 
 const (
 	EventReceivedName = "event/event.received"
 	FnFailedName      = "inngest/function.failed"
+	FnCompletedName   = "inngest/function.completed"
```
Comment (jpwilliams) [inline pkg/event/event.go:17]: This should probably be named `inngest/function.finished` to be inkeeping with other terminology.

This also doesn't fit nicely with `inngest/function.failed`. The downside is that we'd have to create two pauses when waiting for invocation results if we had the two events being sent. My lean is that we ditch `inngest/function.failed` and consolidate everything into the `inngest/function.finished` event, though this would likely break some user flows if they have manually targetted that event.

File: pkg/execution/runner/runner.go [hunk 5]
```diff
@@ -352,7 +352,15 @@ func FindInvokedFunction(ctx context.Context, tracked event.TrackedEvent, fl cqr
...
-	name := evt.Data[consts.InvokeSlugKey]
+	name := ""
+	if inngestObj, ok := evt.Data[consts.InngestEventDataPrefix].(map[string]any); ok {
+		if slug, ok := inngestObj[consts.InvokeFnID].(string); ok {
+			logger.From(ctx).Debug().Str("event", evt.Name).Str("FunctionID", slug).Msg("invoking function by event; it was an invoke event")
+			name = slug
+		} else {
+			return nil, fmt.Errorf("invalid slug type")
+		}
+	}
```
Comment (jpwilliams) [inline pkg/execution/runner/runner.go:363]: 🤮 I'm guessing it's better to create a struct for this instead of manually stepping over `map[string]any`s.

File: pkg/consts/consts.go [hunk 6]
```diff
@@ -64,12 +64,14 @@ const (
...
 	// InvokeEventName is the event name used to invoke specific functions via an
 	// API.  Note that invoking functions still sends an event in the usual manner.
 	InvokeEventName = "inngest/function.invoked"
 	// InvokeSlugKey is the data key used to store the fn name when invoking a function
 	// via an RPC-like call, abstracting event-driven fanout.
-	InvokeSlugKey = "_inngest_fn"
+	InvokeFnID          = "fn_id"
+	InvokeCorrelationId = "correlation_id"
```
Comment (tonyhb) [inline pkg/consts/consts.go:74]: Good ol' MQ.  It's good to stick to the standards — great idea.

File: pkg/devserver/lifecycle.go [hunk 7]
```diff
@@ -45,3 +53,60 @@ func (l lifecycle) OnFunctionScheduled(
...
+
+func (l lifecycle) OnFunctionFinished(
+	ctx context.Context,
+	id state.Identifier,
+	item queue.Item,
+	resp state.DriverResponse,
+	s state.State,
+) {
+	now := time.Now()
+	data := map[string]interface{}{
```
Comment (tonyhb) [inline pkg/devserver/lifecycle.go:65]: FYI, and we do _not_ need to change it, Go 1.20 adds `any` as a nicer keyword for `interface{}`:  `map[string]any`.

File: pkg/devserver/lifecycle.go (hunk 2 above)
Comment (tonyhb) [reply pkg/devserver/lifecycle.go:93]: This is a really good question.  To be honest, I quite like the idea of being able to `POST` the entire event, excluding name:
```
{
  id: "whatever", // for idempotency,
  data: {},
  user: {},
}
```

WDYT?  We might want to be able to invoke with idempotency.  User could be a good idea to add for our future plans.  I don't want to complicate the SDK's API, though.

File: pkg/devserver/lifecycle.go (hunk 3 above)
Comment (tonyhb) [reply pkg/devserver/lifecycle.go:111]: We definitely need to log a hard error here.  Realistically, the one component that _cannot_ go down are our message streams.  Without those, things are in dire straights (as they are for every large company).

File: pkg/event/event.go (hunk 4 above)
Comment (tonyhb) [reply pkg/event/event.go:17]: 100% on `function.finished`.  Ditching `inngest/function.failed` is backcompat difficult.  A migration plan:

1. We start sending one event now: `inngest/function.completed` with a status of success/error/cancel, etc. while keeping `inngest/function.failed.
2. We update the SDK's function handlers to listen to these events with an expression
3. Eventually we deprecate `inngest/function.failed`

That said, failure handling is _so common_ that I don't hate special casing it.

File: pkg/execution/history/lifecycle.go [hunk 8]
```diff
@@ -482,6 +483,102 @@ func (l lifecycle) OnWaitForEventResumed(
 	}
 }
 
+// OnInvokeFunction is called when a function is invoked from a step.
+func (l lifecycle) OnInvokeFunction(
```
Comment (tonyhb) [inline pkg/execution/history/lifecycle.go:487]: change to `OnInvokeScheduled` to fit the rest of the naming params?

File: pkg/execution/history/lifecycle.go [hunk 9]
```diff
@@ -482,6 +483,102 @@ func (l lifecycle) OnWaitForEventResumed(
...
+		if err := d.Write(context.WithoutCancel(ctx), h); err != nil {
+			l.log.Error("execution lifecycle error", "lifecycle", "onInvokeFunction", "error", err)
+		}
+	}
+}
+
+// OnInvokeFunctionResumed is called when a function is resumed from an
+// invoke function step. This happens when the invoked function has
+// completed or the step timed out whilst waiting.
+func (l lifecycle) OnInvokeFunctionResumed(
```
Comment (tonyhb) [inline pkg/execution/history/lifecycle.go:537]: `OnInvokeCompleted`?

File: pkg/execution/runner/runner.go (hunk 5 above)
Comment (tonyhb) [reply pkg/execution/runner/runner.go:363]: 100%, the problem being we can't get that from a plain ol' event.

File: ui/packages/components/src/utils/historyParser/types.ts [hunk 10]
```diff
@@ -37,7 +37,8 @@ export type HistoryType =
   | 'StepScheduled'
   | 'StepSleeping'
   | 'StepStarted'
-  | 'StepWaiting';
+  | 'StepWaiting'
+  | 'StepInvokingFunction';
```
Comment (tonyhb) [inline ui/packages/components/src/utils/historyParser/types.ts:41]: `StepInvoked`?  shorter == better all round.  At scale (and we do hundreds of millions a month) bandwidth costs go up for us and our users.

**Review by tonyhb — COMMENTED** [review]
This is looking really good.

File: pkg/devserver/lifecycle.go (hunk 2 above)
Comment (jpwilliams) [reply pkg/devserver/lifecycle.go:93]: We have support for `data`, `user`, and `v` right now, but `id` is omitted from any event sending in TS. Sounds like we can add that to all entry points (`.invoke()`, `.send()`, `.sendEvent()`) in a follow-up.

File: pkg/devserver/lifecycle.go (hunk 3 above)
Comment (jpwilliams) [reply pkg/devserver/lifecycle.go:111]: Yeah. Let's keep the lifecycle contract the same and I'll pull it out into another function.

File: pkg/consts/consts.go (hunk 6 above)
Comment (jpwilliams) [reply pkg/consts/consts.go:74]: Mostly as we can have many invocations from one run, so anything we have in the triggering event (event ID, run ID) still needs more data adding to it to be meaningful.

File: pkg/execution/history/lifecycle.go (hunk 8 above)
Comment (jpwilliams) [reply pkg/execution/history/lifecycle.go:487]: Matching `OnWaitForEvent`, where we seem to have `On[Tool]`:
```diff
OnWaitForEvent
OnWaitForEventResumed
OnSleep
+ OnInvokeFunction
+ OnInvokeFunctionResumed
```

File: pkg/execution/history/lifecycle.go (hunk 9 above)
Comment (jpwilliams) [reply pkg/execution/history/lifecycle.go:537]: Matching `OnWaitForEvent`, where we seem to have `On[Tool]`:
```diff
OnWaitForEvent
OnWaitForEventResumed
OnSleep
+ OnInvokeFunction
+ OnInvokeFunctionResumed
```

File: ui/packages/components/src/utils/historyParser/types.ts (hunk 10 above)
Comment (jpwilliams) [reply ui/packages/components/src/utils/historyParser/types.ts:41]: I can change the enum: https://github.com/inngest/inngest/blob/571accdd55e046fa18cf311c1e2c2e77742970c1/pkg/enums/history_type.go#L7-L26
Honestly I figured they were serialized as integers. 😅

File: pkg/execution/driver/driver.go [hunk 11]
```diff
@@ -64,8 +73,34 @@ func MarshalV1(
...
+		}
+
+		if len(fn.Triggers) != 1 {
+			return nil, fmt.Errorf("invocation event found, but function %s has %d triggers; could not fill in invocation event name automatically", fn.Slug, len(fn.Triggers))
+		}
+
+		trigger := fn.Triggers[0]
+		// Crons can keep the original event name
+		if trigger.EventTrigger != nil && trigger.Event != "" {
+			evt["name"] = trigger.Event
```
Comment (tonyhb) [inline pkg/execution/driver/driver.go:102]: So, I get it but I'm wondering why this is necessary for the SDK?

A few things that could go wrong:

- It might make external logging lie.  We didn't actually receive this event name.  On the user's side, they'll see many more events of this type than they're actually sending.
- The events received from the SDK won't correlate with our dashboard

Comment (jpwilliams) [reply pkg/execution/driver/driver.go:102]: @tonyhb Two paths here: we can do this or make `event.name` in the SDK be a union of `"the-actual-event" | "inngest/function.invoked"`.

I don't have a strong opinion on this either way - the latter might feel a touch confusing if a user doesn't know what invocation is, but it also provides a discovery path.

I'll delete this code and the SDK can declare it.

File: pkg/execution/history/lifecycle.go [hunk 12]
```diff
@@ -486,6 +487,126 @@ func (l lifecycle) OnWaitForEventResumed(
...
+		Type:            enums.HistoryTypeStepInvokingFunction.String(),
+		Attempt:         int64(item.Attempt),
+		IdempotencyKey:  id.IdempotencyKey(),
+		EventID:         id.EventID,
+		BatchID:         id.BatchID,
+		StepName:        &op.Name,
+		StepID:          &op.ID,
+		InvokeFunction: &InvokeFunction{
+			EventID:       eventID,
+			FunctionID:    fnID,
```
Comment (goodoldneon) [inline pkg/execution/history/lifecycle.go:549]: `fnId` and `expiry` could have their zero values here. Should we set `InvokeFunction` to `nil` in that case? Or should `InvokeFunction.FunctionID` and `InvokeFunction.Timeout` be nullable?

File: pkg/execution/history/lifecycle.go [hunk 13]
```diff
@@ -486,6 +487,126 @@ func (l lifecycle) OnWaitForEventResumed(
...
+			l.log.Error("error parsing invoke function options expiry", "error", err)
+		} else {
+			expiry = optsExp
+		}
+	} else {
+		l.log.Error("invoke function options are nil")
+	}
+
+	h := History{
+		ID:              ulid.MustNew(ulid.Now(), rand.Reader),
```
Comment (goodoldneon) [inline pkg/execution/history/lifecycle.go:532]: Nit: Can you alphabetize the fields? It makes it easier to see if we're missing one, which is tough when there are tons of them. I need to put up a PR to alphabetize the preexisting ones too

File: pkg/execution/history/lifecycle.go (hunk 12 above)
Comment (jpwilliams) [reply pkg/execution/history/lifecycle.go:549]: @goodoldneon Mm, I guess if `fnID` and `expiry` are zero values here then we should just return early instead of logging an error and continuing; they're required to be defined if we're calling this lifecycle hook.

File: pkg/execution/history/lifecycle.go (hunk 13 above)
Comment (jpwilliams) [reply pkg/execution/history/lifecycle.go:532]: @goodoldneon Is there something that does that for me? 😬

Comment (goodoldneon) [reply pkg/execution/history/lifecycle.go:532]: I use vim's `:sort` but VS Code has an equivalent!
image

File: pkg/execution/history/lifecycle.go (hunk 12 above)
Comment (goodoldneon) [reply pkg/execution/history/lifecycle.go:549]: Maybe we should still write but set `InvokeFunction: nil` if we don't have all the data for `InvokeFunction`. Seeing the field as `null` in BigQuery would help troubleshooting

Comment (jpwilliams) [reply pkg/execution/history/lifecycle.go:549]: @goodoldneon I'll add a check that we have all of the data before writing the log.

For the future, is it better for history in these cases to provide only everything/nothing, or is partial data also helpful? It'd lower the quality of the data overall, requiring readers to do more work, though in terms of helping debugging it might provide more threads to pull.

File: pkg/devserver/devserver.go [hunk 14]
```diff
@@ -252,3 +267,84 @@ func createInmemoryRedis(ctx context.Context) (rueidis.Client, error) {
...
+					logger.From(ctx).Debug().Str("data_value_str", dataValue).Msg("data_value")
+					data[consts.InvokeCorrelationId] = dataValue
+				}
+			}
+		}
+
+		if opts.Err != nil {
+			data["error"] = opts.Err
+		} else {
+			data["result"] = opts.Result
```
Comment (tonyhb) [inline pkg/devserver/devserver.go:317]: This seems like it's repeated here and on our cloud.  Is there a way we can normalize this into one function which we use both here and imported into our cloud repo?

File: pkg/execution/executor/executor.go [hunk 15]
```diff
@@ -806,9 +838,33 @@ func (e *executor) HandlePauses(ctx context.Context, iter state.PauseIterator, e
...
 			}
 
+			withEvt := evt.GetEvent()
+			with := withEvt.Map()
+			var runID *ulid.ULID
+
+			isInvokeFunctionOpcode := pause.Opcode != nil && *pause.Opcode == enums.OpcodeInvokeFunction.String()
+			isFnFinishedEvent := withEvt.Name == event.FnFinishedName
+
+			if isInvokeFunctionOpcode && isFnFinishedEvent {
```
Comment (tonyhb) [inline pkg/execution/executor/executor.go:848]: Can we add a comment or extract this into a function so that it's clearer _why_ we need to do this?

On reading the code, it's not obvious that we need to do this specifically because "function invocation should resume with the function's returned data, not with the event wrapper".

File: pkg/execution/executor/executor.go [hunk 16]
```diff
@@ -1168,6 +1232,101 @@ func (e *executor) handleGeneratorSleep(ctx context.Context, gen state.Generator
...
+	// Always create an invocation event.
+	evt := event.NewInvocationEvent(event.NewInvocationEventOpts{
+		Event:         *opts.Payload,
+		FnID:          opts.FunctionID,
+		CorrelationID: &correlationID,
+	})
+
+	logger.From(ctx).Debug().Interface("evt", evt).Str("gen.ID", gen.ID).Msg("created invocation event")
+
+	if e.handleSendingEvent == nil {
```
Comment (tonyhb) [inline pkg/execution/executor/executor.go:1313]: We should check this at the top before starting this function.  Fail early before any mutations.

File: pkg/execution/executor/executor.go [hunk 17]
```diff
@@ -1229,7 +1390,9 @@ func (e *executor) handleGeneratorWaitForEvent(ctx context.Context, gen state.Ge
 	// the pause so this race will conclude by calling the function once, as only
 	// one thread can lease and consume a pause;  the other will find that the
 	// pause is no longer available and return.
+	jobID := fmt.Sprintf("%s-%s-%s", item.Identifier.IdempotencyKey(), gen.ID, "wait")
 	err = e.queue.Enqueue(ctx, queue.Item{
+		JobID:       &jobID,
```
Comment (tonyhb) [inline pkg/execution/executor/executor.go:1395]: Why was this added?  I understand it, but an curious given waitForEvent has existed since day zero.

File: pkg/execution/executor/service.go [hunk 18]
```diff
@@ -114,47 +116,110 @@ func (s *svc) Executor() execution.Executor {
...
+				if err != nil {
+					return fmt.Errorf("error publishing failure event: %w", err)
+				}
+
+				return nil
+			})
 		}
 
-		return nil
+		// send inngest/function.finished
```
Comment (tonyhb) [inline pkg/execution/executor/service.go:168]: Are we doing this on _every_ function finish now?  Or only if there's a correlation ID?

Additionally, should the "function finished" event generation be bundled into the executor itself?  In the cloud we obviously have a wrapper such, but the actual generation of the `event.Event` is the same, right?

**Review by tonyhb — CHANGES_REQUESTED** [review]

File: pkg/execution/executor/service.go (hunk 18 above)
Comment (jpwilliams) [reply pkg/execution/executor/service.go:168]: @tonyhb That's the plan atm from https://github.com/inngest/inngest/pull/782#discussion_r1376256525. We can limit, though an `inngest/function.finished` event is pretty cool to be able to listen to across the system.

It doubles the number of events, of course...

File: pkg/execution/executor/executor.go (hunk 17 above)
Comment (jpwilliams) [reply pkg/execution/executor/executor.go:1395]: Hm I think this might have been creating some duplicate calls after a `waitForEvent` after I'd added `invoke`. https://github.com/inngest/inngest/commit/30c7b8db4bc98914b6d7374b9409c3868c0f8d1b

I can pull this out and debug a touch - a comment would be good if it's the case.

Comment (jpwilliams) [reply pkg/execution/executor/executor.go:1438]: @tonyhb atm, because `waitForEvent` has no `JobID`, it looks like it uses an empty string `""` for idempotency. When that happens, it looks like we instead use a generated ID for idempotency, which is obviously different every time.

If two tools lack setting a `JobID` then parallelism can cause us to re-enqueue the same steps again and again because idempotency is basically removed.

In fact, while this is mostly a fix for adding a new tool, this would definitely cause problems with parallelism and `waitForEvent` anyway. Setting this now won't cause any more damage than it being unset currently. I'll raise a separate PR to get that fix in.

Comment (jpwilliams) [reply pkg/execution/executor/executor.go:1438]: @tonyhb Have added #859 for this.

File: pkg/devserver/devserver.go (hunk 14 above)
Comment (tonyhb) [reply pkg/devserver/devserver.go:324]: I still see this logic - specifically from line 305 -> 325 - duplicated between here and the cloud.  Ideally we'd extract this into a function such that the actual _event generation itself_ is written once and imported.

File: pkg/execution/state/pause.go [hunk 19]
```diff
@@ -160,3 +166,39 @@ func (p Pause) Edge() inngest.Edge {
 		Incoming: p.Incoming,
 	}
 }
+
+type ResumeData struct {
+	// RunID is the
```
Comment (tonyhb) [inline pkg/execution/state/pause.go:171]: Im dying with anticipation :D
